File: frontend/src/types/patent_intelligence.ts
export interface PatentTrendPoint {
  year: number;
  filings_count: number;
  grants_count: number;
  filing_growth_rate: number | null;
}

export interface PatentTrendsResponse {
  total_patents: number;
  year_range: {
    min_year: number | null;
    max_year: number | null;
  };
  points: PatentTrendPoint[];
}

export interface TechnologyDomainItem {
  domain: string;
  patent_count: number;
  share_percentage: number;
  total_citations: number;
  average_citations: number;
  top_assignees: string[];
}

export interface TechnologyDomainsResponse {
  total_domains: number;
  concentration_index_hhi: number;
  concentration_level: 'DIVERSIFIED' | 'MODERATELY_CONCENTRATED' | 'HIGHLY_CONCENTRATED';
  domains: TechnologyDomainItem[];
}

export interface AssigneeLandscapeItem {
  assignee: string;
  patent_count: number;
  share_percentage: number;
  total_citations: number;
  average_citations: number;
  recent_filings_count: number;
  primary_domains: string[];
  jurisdictions: string[];
}

export interface AssigneesResponse {
  total_assignees: number;
  assignee_concentration_hhi?: number;
  concentration_level?: 'DIVERSIFIED' | 'MODERATELY_CONCENTRATED' | 'HIGHLY_CONCENTRATED';
  assignees: AssigneeLandscapeItem[];
}

export interface JurisdictionItem {
  jurisdiction_code: string;
  jurisdiction_name: string;
  patent_count: number;
  share_percentage: number;
}

export interface JurisdictionsResponse {
  total_jurisdictions: number;
  jurisdictions: JurisdictionItem[];
}

export interface PatentStatusItem {
  status: string;
  count: number;
  share_percentage: number;
}

export interface PatentStatusResponse {
  total_patents: number;
  statuses: PatentStatusItem[];
}

export interface CompetitiveAssigneeItem {
  assignee: string;
  patent_count: number;
  domain_breadth_count: number;
  recent_filing_velocity_pct: number;
  average_citations: number;
  competitive_index: number;
  classification: 'DOMINANT_PORTFOLIO' | 'HIGH_VELOCITY' | 'NICHE_SPECIALIST' | 'EMERGING_APPLICANT';
  primary_domains: string[];
  key_drivers: string[];
}

export interface CompetitiveLandscapeResponse {
  total_competitors: number;
  assignee_concentration_hhi?: number;
  concentration_level?: 'DIVERSIFIED' | 'MODERATELY_CONCENTRATED' | 'HIGHLY_CONCENTRATED';
  weighting_schema?: Record<string, number>;
  methodology_disclaimer: string;
  competitors: CompetitiveAssigneeItem[];
}

export interface PatentLandscapeSummary {
  total_patents: number;
  total_assignees: number;
  total_domains: number;
  total_citations: number;
  average_citations: number;
  filing_year_range: {
    min_year: number | null;
    max_year: number | null;
  };
  top_domains: TechnologyDomainItem[];
  top_assignees: AssigneeLandscapeItem[];
  jurisdiction_distribution: JurisdictionItem[];
}

// --- Machine Learning Clustering Types ---
export interface PatentClusterMember {
  patent_id: number;
  patent_number: string;
  title: string;
  assignee: string | null;
  technology_domain: string | null;
  patent_classification: string | null;
  citation_count: number;
  distance_to_centroid: number | null;
  explanation: string;
}

export interface PatentClusterItem {
  cluster_id: number;
  cluster_name: string;
  technology_domain: string;
  patent_count: number;
  share_percentage: number;
  dominant_terms: string[];
  average_citations: number;
  representative_patents: PatentClusterMember[];
  description: string;
}

export interface PatentClusteringResponse {
  total_patents: number;
  total_clusters: number;
  algorithm: string;
  k_requested: number | null;
  clusters: PatentClusterItem[];
  disclaimer: string;
}

// --- Innovation Mapping Types ---
export interface InnovationMapMatrixCell {
  domain: string;
  assignee: string;
  patent_count: number;
  patent_numbers: string[];
}

export interface InnovationMapHotspot {
  domain: string;
  classification: string;
  patent_count: number;
  recent_count: number;
  velocity_score: number;
  top_assignees: string[];
  activity_type: 'EXPANDING_CORE' | 'HIGH_GROWTH' | 'EMERGING';
}

export interface InnovationMapWhitespace {
  domain: string;
  whitespace_reason: string;
  opportunity_level: 'HIGH' | 'MEDIUM' | 'MODERATE';
  description: string;
}

export interface InnovationMapResponse {
  total_patents: number;
  domains: string[];
  assignees: string[];
  classifications: string[];
  matrix: InnovationMapMatrixCell[];
  hotspots: InnovationMapHotspot[];
  whitespaces: InnovationMapWhitespace[];
}

// --- Profile Patent Recommendations Types ---
export interface PatentRecommendationItem {
  patent_id: number;
  patent_number: string;
  title: string;
  abstract: string | null;
  assignee: string | null;
  technology_domain: string | null;
  patent_classification: string | null;
  citation_count: number;
  match_score: number;
  matched_domains: string[];
  matched_keywords: string[];
  rationale: string;
}

export interface PatentRecommendationsResponse {
  total_recommendations: number;
  profile_domains: string[];
  profile_keywords: string[];
  recommendations: PatentRecommendationItem[];
}

File: backend/app/schemas/patent_intelligence.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


# --- Patent Trends ---
class PatentTrendPoint(BaseModel):
    year: int
    filings_count: int
    grants_count: int
    filing_growth_rate: Optional[float] = Field(None, description="Year-over-Year filing growth percentage")


class PatentTrendsResponse(BaseModel):
    total_patents: int
    year_range: Dict[str, Optional[int]]
    points: List[PatentTrendPoint]


# --- Technology Domains & Concentration ---
class TechnologyDomainItem(BaseModel):
    domain: str
    patent_count: int
    share_percentage: float
    total_citations: int
    average_citations: float
    top_assignees: List[str] = []


class TechnologyDomainsResponse(BaseModel):
    total_domains: int
    concentration_index_hhi: float = Field(..., description="Herfindahl-Hirschman Index (0-10000)")
    concentration_level: str = Field(..., description="DIVERSIFIED | MODERATELY_CONCENTRATED | HIGHLY_CONCENTRATED")
    domains: List[TechnologyDomainItem]


# --- Assignees & Top Patent Holders ---
class AssigneeLandscapeItem(BaseModel):
    assignee: str
    patent_count: int
    share_percentage: float
    total_citations: int
    average_citations: float
    recent_filings_count: int
    primary_domains: List[str] = []
    jurisdictions: List[str] = []


class AssigneesResponse(BaseModel):
    total_assignees: int
    assignee_concentration_hhi: float = Field(0.0, description="Herfindahl-Hirschman Index for Assignee concentration (0-10000)")
    concentration_level: str = Field("DIVERSIFIED", description="DIVERSIFIED | MODERATELY_CONCENTRATED | HIGHLY_CONCENTRATED")
    assignees: List[AssigneeLandscapeItem]


# --- Jurisdictions ---
class JurisdictionItem(BaseModel):
    jurisdiction_code: str
    jurisdiction_name: str
    patent_count: int
    share_percentage: float


class JurisdictionsResponse(BaseModel):
    total_jurisdictions: int
    jurisdictions: List[JurisdictionItem]


# --- Patent Status ---
class PatentStatusItem(BaseModel):
    status: str
    count: int
    share_percentage: float


class PatentStatusResponse(BaseModel):
    total_patents: int
    statuses: List[PatentStatusItem]


# --- Competitive Landscape Indicators ---
class CompetitiveAssigneeItem(BaseModel):
    assignee: str
    patent_count: int
    domain_breadth_count: int
    recent_filing_velocity_pct: float
    average_citations: float
    competitive_index: float = Field(..., ge=0.0, le=100.0)
    classification: str = Field(..., description="DOMINANT_PORTFOLIO | HIGH_VELOCITY | NICHE_SPECIALIST | EMERGING_APPLICANT")
    primary_domains: List[str] = []
    key_drivers: List[str] = []


class CompetitiveLandscapeResponse(BaseModel):
    total_competitors: int
    assignee_concentration_hhi: float = Field(0.0, description="Herfindahl-Hirschman Index for Assignee concentration (0-10000)")
    concentration_level: str = Field("DIVERSIFIED", description="DIVERSIFIED | MODERATELY_CONCENTRATED | HIGHLY_CONCENTRATED")
    weighting_schema: Dict[str, float] = Field(
        default_factory=lambda: {"volume": 0.40, "velocity": 0.30, "citations": 0.20, "domain_breadth": 0.10},
        description="Weighting breakdown for Composite Competitive Index"
    )
    methodology_disclaimer: str = Field(
        "Structured Patent Landscape Indicator based on patent metadata and filing activity. Does not represent full commercial market viability.",
        description="Transparency disclaimer"
    )
    competitors: List[CompetitiveAssigneeItem]


# --- Comprehensive Landscape Overview ---
class PatentLandscapeSummary(BaseModel):
    total_patents: int
    total_assignees: int
    total_domains: int
    total_citations: int
    average_citations: float
    filing_year_range: Dict[str, Optional[int]]
    top_domains: List[TechnologyDomainItem]
    top_assignees: List[AssigneeLandscapeItem]
    jurisdiction_distribution: List[JurisdictionItem]


# --- Real Machine Learning Patent Clustering ---
class PatentClusterMember(BaseModel):
    patent_id: int
    patent_number: str
    title: str
    assignee: Optional[str] = None
    technology_domain: Optional[str] = None
    patent_classification: Optional[str] = None
    citation_count: int = 0
    distance_to_centroid: Optional[float] = None
    explanation: str


class PatentClusterItem(BaseModel):
    cluster_id: int
    cluster_name: str
    technology_domain: str
    patent_count: int
    share_percentage: float
    dominant_terms: List[str]
    average_citations: float
    representative_patents: List[PatentClusterMember]
    description: str


class PatentClusteringResponse(BaseModel):
    total_patents: int
    total_clusters: int
    algorithm: str = Field("TF-IDF Vectorization + K-Means Clustering", description="Applied ML clustering algorithm")
    k_requested: Optional[int] = None
    clusters: List[PatentClusterItem]
    disclaimer: str = "Unsupervised machine learning clustering based on scikit-learn TF-IDF text representations and K-Means centroid optimization."


# --- Innovation Mapping ---
class InnovationMapMatrixCell(BaseModel):
    domain: str
    assignee: str
    patent_count: int
    patent_numbers: List[str] = []


class InnovationMapHotspot(BaseModel):
    domain: str
    classification: str
    patent_count: int
    recent_count: int
    velocity_score: float
    top_assignees: List[str] = []
    activity_type: str = "EXPANDING_CORE"  # EXPANDING_CORE | HIGH_GROWTH | EMERGING


class InnovationMapWhitespace(BaseModel):
    domain: str
    whitespace_reason: str
    opportunity_level: str = "HIGH"  # HIGH | MEDIUM | MODERATE
    description: str


class InnovationMapResponse(BaseModel):
    total_patents: int
    domains: List[str]
    assignees: List[str]
    classifications: List[str]
    matrix: List[InnovationMapMatrixCell]
    hotspots: List[InnovationMapHotspot]
    whitespaces: List[InnovationMapWhitespace]


# --- Module 3 Profile-Based Patent Recommendations ---
class PatentRecommendationItem(BaseModel):
    patent_id: int
    patent_number: str
    title: str
    abstract: Optional[str] = None
    assignee: Optional[str] = None
    technology_domain: Optional[str] = None
    patent_classification: Optional[str] = None
    citation_count: int = 0
    match_score: float = Field(..., ge=0.0, le=100.0)
    matched_domains: List[str] = []
    matched_keywords: List[str] = []
    rationale: str


class PatentRecommendationsResponse(BaseModel):
    total_recommendations: int
    profile_domains: List[str] = []
    profile_keywords: List[str] = []
    recommendations: List[PatentRecommendationItem]
